### PR TITLE
SNOW-1435047 fix insert with binding data on different paramstyle

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1435047

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 - Improved error message to remind users set `{"infer_schema": True}` when reading csv file without specifying its schema.
 
+### Bug Fixes
+
+- Fixed a bug where `Session.create_dataframe` when created with large greater than 512 rows threw `TypeError` when created a session with `format` or `pyformat` `paramstyle`.
```
Comment (sfc-gh-jrose) [inline CHANGELOG.md:11]:
```suggestion
- Fixed a bug where `Session.create_dataframe` is called with more than 512 rows throws `TypeError` when creating a session with `format` or `pyformat` `paramstyle`.
```

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 2]
```diff
@@ -563,7 +563,9 @@ def large_local_relation_plan(
             is_generated=True,
         )
         insert_stmt = batch_insert_into_statement(
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:565]: Is this function ever called anywhere else? If so maybe make the paramstyle have a default value that matches the old value?

**Review by sfc-gh-jrose — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:565]: Its only called once here

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,10 @@
 
 - Improved error message to remind users set `{"infer_schema": True}` when reading csv file without specifying its schema.
 
+### Bug Fixes
+
+- Fixed a bug where `Session.create_dataframe` is called with more than 512 rows throws `TypeError` when creating a session with `format` or `pyformat` `paramstyle`.
```
Comment (sfc-gh-aling) [inline CHANGELOG.md:11]: just reading the code, why it's the number `512` that triggers the error?

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 4]
```diff
@@ -774,12 +776,22 @@ def insert_into_statement(
...
+def batch_insert_into_statement(
+    table_name: str, column_names: List[str], paramstyle: str
+) -> str:
+    num_cols = len(column_names)
+    if paramstyle == "qmark":
+        placeholder_marks = [QUESTION_MARK] * num_cols
+    elif paramstyle == "numeric":
+        placeholder_marks = [f"{SINGLE_COLON}{i+1}" for i in range(num_cols)]
+    else:  # for pyformat and format
+        placeholder_marks = [PERCENT_S] * num_cols
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:788]: It seems like we previously didn't support format/pyformat/numeric binding at all.. should this be considered a new feature?

File: CHANGELOG.md (hunk 3 above)
Comment (sfc-gh-aalam) [reply CHANGELOG.md:11]: 512 is set as `analyzer.ARRAY_BIND_THRESHOLD`

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py (hunk 4 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:788]: sure. I don't mind

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 5]
```diff
@@ -774,12 +776,22 @@ def insert_into_statement(
...
-def batch_insert_into_statement(table_name: str, column_names: List[str]) -> str:
+def batch_insert_into_statement(
+    table_name: str, column_names: List[str], paramstyle: str
+) -> str:
+    num_cols = len(column_names)
+    if paramstyle == "qmark":
+        placeholder_marks = [QUESTION_MARK] * num_cols
+    elif paramstyle == "numeric":
+        placeholder_marks = [f"{SINGLE_COLON}{i+1}" for i in range(num_cols)]
+    else:  # for pyformat and format
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py]: good suggestion. updated the PR.

**Review by sfc-gh-dpetersohn — APPROVED** [review]
Thanks, looks great!

**Pull request merged**